### PR TITLE
Improvements to AnimationNodeStateMachine

### LINKED ISSUE #19773: [Proposal StateMachine] Add condition to transitions
**Godot version:**
8c7da84e1eae25302a0c88594cce88b84f2dc21e

**OS/device including version:**
Any

**Issue description:**

The current state machine doesn't have conditions for the transitions, so it is not possible to
control the transition of states from "environment" variables such as Unity or Unreal.

Unreal: https://docs.unrealengine.com/en-us/Engine/Animation/StateMachines/TransitionRules
Unity: https://docs.unity3d.com/Manual/StateMachineTransitions.html

My proposal is implemeting something simple, similar to the Unity.

Here is a prototype of what I was trying to do in godot:
captura de tela de 2018-06-25 17-33-19
*The final version will be much bigger*

Example of what I want to create in godot:
captura de tela de 2018-06-25 19-49-12

I've created my own state machine entirely on gdscript for my project and it looks like this:
```
# player.gd

onready var state_machine = preload("res://scripts/state_machine.gd").new($animation_player)

func _physics_process(delta):
	_apply_movement(delta)
	_apply_animation()

func _apply_movement(delta):
	# All movement code goes here

func _apply_animation():
	# Here I just need to pass some variables to state machine,
	state_machine.lv = lv
	state_machine.left = left
	state_machine.right = right
	state_machine.down = down
	state_machine.is_on_floor = is_on_floor()
	state_machine.speed = SPEED
	state_machine.speed_run = SPEED_RUN
	state_machine.jump = jump
	state_machine.shooting = shoot
	state_machine.ledge_climbing = ledge_climbing
	state_machine.ledge_hanging = ledge_hanging
	state_machine.aim = aim
	state_machine.turn = false

	# All logic of my state machine will run here.
	# After the update(), my AnimationPlayer will play the correct animation.
	state_machine.update()
```
And I expect that node "StateMachine" works in a similar way.

The benefits of this IMO are that "movement" code gets simpler because I don't have to worry about which animation I should play (and in my case, I have so many states that it's very difficult to work with it without a state machine) and it's more easy to debug (my current SM is a bunch of code and I need to use [draw.io](https://www.draw.io/) to help me visualize when I'm making the states/transitions).

So...
1) We'll need a way to specify variables when creating the state machine to later define these variables by code (like the example code above).
2) All transitions will need a condition that will use these variables that we define to control when you need to move to another state.
Real example of some transition condition to "Idle" state on my project: `is_on_floor and abs(lv.x) == 0`.
3) I think it would be useful a way to see all conditions from incoming transitions just clicking in the state. Something like this: 
captura de tela de 2018-06-26 10-56-50

4) ~Maybe transitions could be some kind of resource, that way, we can use the same condition several times. eg: Most of my states use the same condition when goes to "Idle".~ If item 3 could be implemented, We can just copy/paste the conditions.
5) Right now, I can't select multiple states (useful to reorganize multiple states).
6) In my game, I've created some "shortcut" states (eg: "Any state", "All aims") just to simplify, I think maybe this could be done in godot using Sub State Machines but I can't test it without the conditions.
Something like this: 
captura de tela de 2018-06-26 11-17-52
*Example of Any State*
Instead of this:
captura de tela de 2018-06-26 11-16-41
7. Also, I liked this options below, especially the "At End". I implemented exactly the same thing on my state machine, this means that I will only accept the condition if the animation of the current state is over. I would like to preserve this behavior.
captura de tela de 2018-07-06 11-11-17

### PR DESCRIPTION
**Pull request by guilhermefelipecgs**
Closes #19773.

Notes about the implementation: https://github.com/godotengine/godot/pull/24402#issuecomment-1038117414

Open the menu to add new animation nodes by dragging the transitions to empty areas and automatically connecting them:
peek 10-02-2019 16-34
Adds box selection to the state machine:
peek 10-02-2019 16-35
Add feature to group/ungroup selected nodes in a "sub" state machine:
peek 10-02-2019 16-37
Add popup to connect nodes in sub state machine:
peek 10-02-2019 16-40
Add new "type" of transition line when multiple transitions are grouped:
peek 10-02-2019 16-44
Add dialog to select which transitions can be deleted when they are grouped:
peek 10-02-2019 16-45

### CODE REVIEW

**Comment by mhilbrunner** [comment]
Whats the state of this? Because of the WIP in the title.

**Comment by guilhermefelipecgs** [comment]
I finished implementing the last method that was missing AnimationNodeStateMachine::_get_shortest_pat` and now it's ready to review.

**Comment by aaronfranke** [comment]
Is there a reason that this needs to be in core and not an add-on in the asset library? There are already [many state machine assets](https://godotengine.org/asset-library/asset?category=&filter=state+machine) available on the asset library.

**Comment by Calinou** [comment]
@aaronfranke This PR is for the animation state machine already present in the editor. This PR doesn't add a generic state machine.

**Comment by guilhermefelipecgs** [comment]
Rebased

**Comment by 14NGiestas** [comment]
Any plans on supporting some kind of global "parameters/conditions"? 
When dealing with multiple sub machines one may want to set a certain "condition_a" that is the same for all sub states.

```gdscript
fsm["parameters/conditions/condition_a"] = (...) # Root
fsm["parameters/SubMachine1/conditions/condition_a"] = (...)
fsm["parameters/SubMachine2/conditions/condition_a"] = (...)
fsm["parameters/SubMachine3/conditions/condition_a"] = (...)
```


It gets nightmarish with more sub machines:

```gdscript
fsm["parameters/conditions/condition_a"] = (...) # Root
fsm["parameters/SubMachine1/InnerState1/conditions/condition_a"] = (...)
fsm["parameters/SubMachine1/InnerState2/conditions/condition_a"] = (...)
fsm["parameters/SubMachine1/InnerState3/conditions/condition_a"] = (...)
fsm["parameters/SubMachine2/InnerState1/conditions/condition_a"] = (...)
fsm["parameters/SubMachine3/InnerState2/conditions/condition_a"] = (...)
(...)
```

With a global condition one can simply write:
```gdscript
fsm["parameters/global_conditions/condition_a"] = (...)
```

---
I have the following use case:

A main state machine with general "Idle" / "Move" / "Jump" / "Fall" cycle

main

and other two sub state machines resources, one dealing with "fire logic" and "crouch-prone-stand logic" as resources so I can easily reuse then by loading in the editor in other needed states (like "Walk").

The crouch prone stand is simple:

main

The firing logic is a little bit more complex:

main5

**Comment by fire** [comment]
What is needed for review from the animation team?

**Comment by aaronfranke** [comment]
@fire Are the changes in this PR the right approach? Is the code good (aside from needing a rebase)?

**Comment by guilhermefelipecgs** [comment]
@14NGiestas sry for the delay, I didn't see your message.

> Any plans on supporting some kind of global "parameters/conditions"?

I don't have plans for that right now, but it can be done easily with gdscript. I'm preparing a demo to showcase the changes it this PR in the hope it can get finally approved. I my demo I basically create a method that find all playbacks so when I need to use global condition (eg. for a die state) I call this method here `_set_param('die', true, true)`.

Here is the code I use:
``` gdscript
var playbacks

func _ready():
	_init_playbacks()


func _init_playbacks():
	for item in $animation_tree.get_property_list():
		if item.name.match('*playback'):
			var path = item.name.replace('playback', '')
			playbacks.append(path)


func _set_param(condition, value, use_all_playbacks = false):
	# set condition for all state machines to false
	for playback in playbacks:
		var param = playback + 'conditions/' + condition
		if $animation_tree.get(param) != null:
			$animation_tree[param] = false

	if use_all_playbacks:
		# set all state machines
		for playback in playbacks:
			var param = playback + 'conditions/' + condition
			if $animation_tree.get(param) != null:
				$animation_tree[param] = value
	else:
		# set only current sub state machine
		var param = _get_current_playback() + 'conditions/' + condition
		if $animation_tree.get(param) != null:
			$animation_tree[param] = value


func _get_current_playback() -> String:
	var sm : AnimationNodeStateMachine = $animation_tree.tree_root
	var current_node = str($animation_tree['parameters/playback'].get_current_node())
	var current_path = '/'

	while sm.has_node(current_node):
		if sm.get_node(current_node) is AnimationNodeStateMachine:
			sm = sm.get_node(current_node)
			current_path += current_node + '/'

			if $animation_tree.get('parameters' + current_path + 'playback'):
				current_node = str($animation_tree.get('parameters' + current_path + 'playback').get_current_node())
		else:
			break

	return 'parameters' + current_path
```

**Comment by SaracenOne** [comment]
The subject of this PR came up in the PR review meeting today. I think people do feel it's very unfortunate that this has been so backlogged for so long, so I wanted to make sure I reviewed this. For the most part, I think this will be fine. The changes here seem useful and convinent.

I suppose it would be nice to have a dedicate button for state machine grouping/ungrouping, but its otherwise a fine QOL improvement when you know it exists. I did notice a minor issue where the transition arrow would sometimes appear to get stuck after dismissing the popup menu for creating new nodes. This wasn't a functional problem, just a visual glitch which looks a little janky. It would be nice if it could be fixed, but given how long this PR has been here, I wouldn't even advocate holding up merging this over something this minor. It can be polished up in a seperate PR.

Also, if you're too busy or need help with rebasing, let me know and I could have a go getting this ready for merging.

**Comment by guilhermefelipecgs** [comment]
Hi @SaracenOne, thanks for the reply. At the moment I'm doing some important bug fixes (not focusing on UI right now) that I found this week (now I just need to fix a problem with the cross fade when exiting a sub state machine) and next week I'm planning to finish the rebase and also post here some demos (with this demos will be easier to test if the rebase it's working) and details about the implementation of this PR that I think is relevant.

The rebase should be simple but if I have any problems or questions I'll call you to help.

**Comment by guilhermefelipecgs** [comment]
Rebase done! Below are some notes about the PR.

# Motivation

My main motivation for opening this PR was not the usability itself but the change in the behavior of the state machine. Without this change it is difficult to create more complex machines. I'm sharing the source code of a project I did (it's not that polished but it's a good example IMO) so people can test/study this new system.

# Implementation details

Most of this PR is just usability but the most important is this change here:

https://github.com/godotengine/godot/blob/f956d15cb0a052455aff38db12e3251bd9bb5096/scene/animation/animation_node_state_machine.h#L157-L169

Before, the from/to variables pointed to the names of states within the current context (state machine) only, and with that it was not possible to connect states on different machines. Now from/to uses the relative path and local_from/local_to method was created as a shortcut to be able to get the local path.

Let's look at an example from the demo:
<details>
  <summary>Click to expand!</summary>
  
Screenshot from 2022-02-11 12-17-46
</details>

The first noticeable difference that we can notice in relation to the old code is that this type of machine would not be possible to be created because this state machine has 4 ways to start/end compared to 1 of the old code and this is possible because now the states directly connect to other states on other machines. Second change that now the start/end is no longer an `AnimationNodeAnimation`, so they no longer have animation length and that makes sense for this new system.

Taking 3 examples of how transitions are recorded internally with relative path:

```
../idle -> jump
../run/run_2_idle/run_2_idle_R_after -> jump
../run/run_cycle -> jump_log
```

Before it would look like this:
```
Start -> jump
Start -> jump
Start -> jump_log
```

But realize that this would not make sense, because the first 2 transitions when converted to the local path become `Start -> jump` and end up being duplicated since it is not possible to make this type of transition today and the `Start` node would have to be an animation.

Also, both transitions has the same `advance_condition` named `jump`, and speaking in a nutshell, with a single condition the state machine knows which state to start thanks to the relative path.

<details>
  <summary>Click to expand!</summary>

Screenshot from 2022-02-12 17-59-47

</details>

# About the state machine demo

The demo has two state machines, one more complex for the character and another simpler for the targets. I tried to do something relatively complex to test as many transition combinations as possible. The image below shows an expanded view of the character's machine (If I were to develop a full version I believe that would be about 30% of the total animations for the character):

<details>
  <summary>Click to expand!</summary>
  
minimap2

</details>

I also made a tool to pause the game and advance frame by frame along with a remote debug prototype so people can test and see in real time what is happening:

<details>
  <summary>Click to expand!</summary>

https://user-images.githubusercontent.com/1387165/153727997-7ee4de7c-46c2-4ea8-a044-ee61bc9079fa.mp4

</details>

# Cross fade demo

This is a simpler project just to test the xfade property and make sure it's working.

<details>
  <summary>Click to expand!</summary>

It has two animations moving in two directions:

https://user-images.githubusercontent.com/1387165/153728652-ce5445fa-e702-4341-9a9e-91c759b8c9ee.mp4

And three state machines with a different structure using xfade at 0.5:

https://user-images.githubusercontent.com/1387165/153728643-f62b9fac-e06b-4aa5-a73a-0cc56a5919c1.mp4
  
</details>

# Compatibility Break

AFAIK you only need to connect start/end nodes to migrate for the new version like this:
<details>
  <summary>Click to expand!</summary>
  
Before:

Screenshot from 2022-02-11 15-45-44

After:

Screenshot from 2022-02-11 15-47-04

</details>

# Some things that could be improved
I intend to work on some of the simpler items in the list below such as bugs and improvements that are simpler to implement but I think we could release the PR in the meantime so people can test it out.
- group/ungroup button
- fix some drawing bugs
- remote debugger (I have no idea how it works to synchronize the game with the editor but it would be an essential feature to debug more complex machines like the one I did)
- remove animation postition preview of start/end node
- minimap
- zoom in/out
- warning when ungrouping machines that have transitions directly connected to the end node as they will be removed in the process
- grid style like `GraphEdit` (but not the align system)
- a better way to see all transition's priority coming out of a specific node
- better inspector for multi transition (I didn't really like the result, I tried to put `from -> to` in place of the inspector's indices but when it has the character `/` creates a subgroup and so I ended up using the numbers and creating a `transition_path` property)
  
  Screenshot from 2022-02-11 12-45-27
- `disable` transitions currently do nothing
- global conditions

# Demo projects
- https://github.com/guilhermefelipecgs/state_machine_demo
- [Test xfade.zip](https://github.com/godotengine/godot/files/8055600/Test.xfade.zip)

**Comment by guilhermefelipecgs** [comment]
PR ready for review again.
- [x] Add group/ungroup button
- [x] Implement disabled transition
- [x] Remove play position from start/end node
- [x] Many visual bugs fixed

File: scene/animation/animation_tree.h
Comment (reduz) [inline scene/animation/animation_tree.h:169]: Maybe they should be StartState and EndState ?

File: scene/animation/animation_node_state_machine.h
Comment (reduz) [inline scene/animation/animation_node_state_machine.h:160]: While good in intention, this will unfortunately incur a _very_ large performance penality as it beats the purpose of using stringnames (which are just a pointer comparison). The local values need to be cached somehow.

**Review by reduz — CHANGES_REQUESTED** [review]

File: scene/animation/animation_tree.h
Comment (guilhermefelipecgs) [reply scene/animation/animation_tree.h]: Classes renamed. I also renamed the `MultiAnimationNodeStateMachineTransition` to `MultiTransitionEdit`

File: scene/animation/animation_node_state_machine.h
Comment (guilhermefelipecgs) [reply scene/animation/animation_node_state_machine.h]: Adjusted.

**Comment by SaracenOne** [comment]
It looks like the issues I mentioned in my last assessment of this have all been addressed. Seems pleasantly clean to work with, and appreciate little touches like the dedicated menu for deleting individual transitions.

Two issues I noticed though: the state machine editor, there are now options to create StartState and EndState which I don't think are meant to created via this menu; they don't seem to do anything, and just give unfriendly error messages when attempting to create them
gU7VJk9ESd
Second is that while disabling transitions seem to work, the transtion line is only greyed out if the first transition in the list is disabled. My gut instinct is it should only be greyed if all the transitions are disabled.

Bear in mind, I have not reviewed the code of this implementation yet, this purely an assessment from a functionality and usability standpoint, in which case, it seems fine.

**Review by SaracenOne — CHANGES_REQUESTED** [review]
A lot of changes, but nothing seems obviously wrong here. Can we stop StartState and EndState from appearing in the right-click menu of StateMachineEditor though?

**Comment by guilhermefelipecgs** [comment]
> Can we stop StartState and EndState from appearing in the right-click menu of StateMachineEditor though?

It was definitely not supposed to appear in the menu, when I renamed it I forgot to test it.

Both issues have been fixed.

**Comment by guilhermefelipecgs** [comment]
Note that the state machine is causing the engine to crash but it was not my fault, this bug is happening in the master too.

**Review by SaracenOne — APPROVED** [review]
Seems good to me now

File: editor/plugins/animation_state_machine_editor.h
Comment (reduz) [inline editor/plugins/animation_state_machine_editor.h:237]: May be better to rename to `EditorAnimationMultiTransitionEdit` for better context

**Review by reduz — CHANGES_REQUESTED** [review]

File: editor/plugins/animation_state_machine_editor.h
Comment (guilhermefelipecgs) [reply editor/plugins/animation_state_machine_editor.h]: Renamed.

**Comment by lyuma** [comment]
@SaracenOne I pushed a rebased version here. Feel free to use this commit as your own.
https://github.com/lyuma/godot/tree/guilhermefelipecgs_state_machine_improvement

so I've been playing with it. I was able to replicate one complex state machine example sent to me by a Unity animation guru so it seems pretty powerful. Here are the issues I ran into:
1. (Not a new issue) transition order is not shown in the UI when going to two separate states so you either have to remember or hand edit the resource. In the .tres file, last transition takes precedence which I can work with but it's opposite of Unity.
2. Still no way to have duplicate transitions from one state to another. This is useful if you have like override variable to go to state 1, then try to go to state 2,3,4 etc. then By default go to state 1 with autoadvance. The easiest way to do this is duplicate state 1.
3. The example I did involved redirecting through multiple Start->End->Start transitions of different state machines same-frame. In Godot, I could see it flicker through these different transitions so it leads me to ask...what happens to cross fading  when traveling from one state machine to another?
( In Unity there's a technical explanation why cross fading still works here: the transitions starting at a state machine, and transitions starting at Start are greyed out to indicate they can only add conditions. These "Transition", not "StateTransition" lines specifically do not affect cross fading, and they do not take an extra frame to evaluate. )
4. Start and End settings got lost when migrating old animation trees. There needs to be a migration path, for example read the start flag and auto create a default transition from Start to x... and same for End. (There's no longer an End in the top level state machine)... also rename any existing states named "Start" or "End". Who knows what users did before.
5. Not your fault but switching between the arrow tool and the transition tool is awkward as ever. I keep getting confused because I'm on the wrong tool and it's tedious to switch. I want like right cliick drag and "Add Transition" shows up in the context menu or have a dot you can drag like other node graphs in Godot
6. Transition conditions are a bit tricky to use since there's no not, but (1) allowed me to workaround it by adding an autoadvance and then the true case as two separate transitions. Reduz has an open PR for this, so no need to fix it.

That's basically it. Overall, the PR works and seems to solve most issues I had last time I tried sub State Machines. I know I listed a lot of things, but the only one which is a blocker for me to use it is the frame delay / no cross fading... that's important to have for a character controller, but otherwise, this PR did most things right.

(Also note that if you take my cross fading suggestion (3), it may be possible to soft-lock with nested state machine transitions. The version of Unity I tried is dreadfully easy to soft lock if you add a loop between start and end of multiple state machines. If you're worried about this, set a limit of how many transitions you will follow at once.)

**Comment by lyuma** [comment]
Ok so I did more testing, and I think my assessment was wrong, and it seems to work fine when using crossfade, at least after changing my example a bit.

Here is the state machine I ended up creating:
image
You can download the scene here with a cube and simple animations embedded: [MultiStateMachineTest.tscn](https://github.com/godotengine/godot/files/8587352/MultiStateMachineTest.tscn.txt)

It's a little absurd, but I structured it like this to match the unity example I was basing this on, and it seemed like a good challenge and a way to learn a bit more how things work.

So the part that actually threw me off is that transitions with xfade_time set seem to become "At End". it's not completely consistent, so it's hard for me to say what's correct.

The Loopback thing was my attempt to avoid needing two copies of Logic (allow transitioning back to my own state machine), so I connected Start->End. Imagine you make a state machine for Jump, and the player holds spacebar so you want to xfade from one jump directly into another jump. Anyway... when doing this, it does seem to have some flickering, but I'm not sure if this case is supposed to work or not.

Overall, I think I'm a little more comfortable with this system than when I started using it. I'm generally confused how xfade_time and "Immediate" are supposed to work together. I'm a little worried that I needed to use xfade both inside the state machine and outside of it, which makes me wonder if it's overlapping two crossfades, since that would cause interpolation between state machines to be non-linear.

What are your thoughts?

**Comment by guilhermefelipecgs** [comment]
@SaracenOne the bug I mentioned was a regression, I opened an issue here https://github.com/godotengine/godot/issues/60710 so if the PR is good I think we can continue with the merge,

**Review by reduz — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks for keeping up with a lengthy review process :)

**Comment by ncollie42** [comment]
This commit makes StateMachine now only usable inside other StateMachine. This breaks compatibility between BlendTrees and StateMachine. A StateMachine created inside a BlendTrees will not be able to exit due to missing end nodes. Is this intentional?

**Pull request merged**
